### ISSUE TITLE
Redesign the way `RandomCoin` works in the context of RPO

### ISSUE THREAD

**Issue by Al-Kindi-0** [0xPolygonMiden/miden-vm#881]
The following is proposal for implementing a Miden VM-tailored `RandomCoin` that is duplex-based for use as a random oracle for Fiat-Shamir.
The main difference between the current implementation and the standard one is related to
the fact that we do not pad before applying the permutation. This makes things a bit cheaper
in Miden VM but should be investigated from a security point of view. If this proves to be
problematic, we can use padding before each permuation at a small cost.

``` rust
impl RpoRandomCoinDuplex {
    pub fn new(seed: &[Felt]) -> Self {
        let mut state = [ZERO; Rpo256::STATE_WIDTH];
        let rate_width = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
        let mut i = 0;
        for &element in seed.iter() {
            state[Rpo256::RATE_RANGE.start + i] = element;
            i += 1;
            if i % rate_width == 0 {
                Rpo256::apply_permutation(&mut state);
                i = 0;
            }
        }
        // Apply one last permutation in case seed.len() is not divisible by rate_width
        Rpo256::apply_permutation(&mut state);

        // Create output buffer
        let mut output = vec![];
        output.extend_from_slice(&state[Rpo256::RATE_RANGE]);

        RpoRandomCoinDuplex { state, output }
    }

    fn draw_basefield(&mut self) -> Felt {
        if self.output.is_empty() {
            Rpo256::apply_permutation(&mut self.state);
            self.output.extend_from_slice(&self.state[Rpo256::RATE_RANGE]);
        }
        self.output.pop().expect("Should not fail given the above")
    }
}

impl RandomCoin for RpoRandomCoinDuplex {
    type BaseField = Felt;

    type Hasher = Rpo256;

    fn new(seed: &[Self::BaseField]) -> Self {
        let mut state = [ZERO; Rpo256::STATE_WIDTH];
        let rate_width = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
        let mut i = 0;
        for &element in seed.iter() {
            state[Rpo256::RATE_RANGE.start + i] = element;
            i += 1;
            if i % rate_width == 0 {
                Rpo256::apply_permutation(&mut state);
                i = 0;
            }
        }
        // Apply one last permutation in case seed.len() is not divisible by rate_width
        Rpo256::apply_permutation(&mut state);

        // Create output buffer
        let mut output = vec![];
        output.extend_from_slice(&state[Rpo256::RATE_RANGE]);

        RpoRandomCoinDuplex { state, output }
    }

    fn reseed(&mut self, data: <Self::Hasher as winter_crypto::Hasher>::Digest) {
        // Clear buffer
        self.output.clear();

        // Overwrite the first half of the rate portion with the digest
        let data = data.as_elements();
        let rate_width = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
        for i in 0..rate_width {
            self.state[Rpo256::RATE_RANGE.start + i] = data[i];
        }

        // Absorb
        Rpo256::apply_permutation(&mut self.state);

        // Update output buffer
        self.output.extend_from_slice(&self.state[Rpo256::RATE_RANGE]);
    }

    fn reseed_with_int(&mut self, value: u64) {
        let value = Felt::new(value);
        self.state[Rpo256::RATE_RANGE.start] = value;
        Rpo256::apply_permutation(&mut self.state);
    }

    fn draw<E: FieldElement<BaseField = Felt>>(&mut self) -> Result<E, RandomCoinError> {
        let ext_degree = E::EXTENSION_DEGREE;
        let mut result = vec![];
        for i in 0..ext_degree {
            result[i] = self.draw_basefield();
        }
        let result = E::slice_from_base_elements(&result);
        Ok(result[0])
    }

    fn leading_zeros(&self) -> u32 {
        todo!()
    }

    fn check_leading_zeros(&self, value: u64) -> u32 {
        todo!()
    }


    fn draw_integers(
        &mut self,
        num_values: usize,
        domain_size: usize,
    ) -> Result<Vec<usize>, RandomCoinError> {
        todo!()
    }
}
```

**Comment by bobbinth** [comment]
Looks good! We can probably apply some small optimizations here and there, but this is pretty much exactly how I would have done it as well. A couple of comments/questions:

1. Should we keep the name just `RpoRandomCoin` (without the `duplex` part)?
2. The constructor in `impl RpoRandomCoinDuplex` is probably not needed, right?
3. When hashing the seed during initialization, I think we should apply proper RPO padding rule. I don't think this will introduce significant overhead in MASM.
4. Is your comment about padding related to the above? Or should we be applying some kind of padding for each "squeezing"?

**Comment by Al-Kindi-0** [comment]
> Looks good! We can probably apply some small optimizations here and there, but this is pretty much exactly how I would have done it as well. A couple of comments/questions:
> 
>     1. Should we keep the name just `RpoRandomCoin` (without the `duplex` part)?
> 
Makes sense, for now using duplex suffix to separate it from the current implementation.
>     2. The constructor in `impl RpoRandomCoinDuplex` is probably not needed, right?
> 
>     3. When hashing the seed during initialization, I think we should apply proper RPO padding rule. I don't think this will introduce significant overhead in MASM.
> 
Makes sense, we can hash the seed using standard RPO (with padding) and then just call reseed. Thus we won't need a constructor.
>     4. Is your comment about padding related to the above? Or should we be applying some kind of padding for each "squeezing"?
It is more about the latter. More specifically, the duplex construction absorbs a number of elements (after padding) that is less than the rate width and squeezes a certain number of elements that is again less than the rate width.  When in add-mode ( the other mode being overwrite-mode), we can generalize this to squeezing a sequence of arbitrary length by imagining that some of the input sequences are zero (and thus we can call permute directly and without erasing the rate-portion of the state). This also allows us to avoid the need for padding by assuming that each absorption step absorbs a sequence of length equal to the rate width by adding zeros to reach it.
The question now is how to make this work in overwrite mode (efficiently). Now that I think about it, we might assume that the duplex works with add-mode and the sponge in overwrite-mode. It remains to be checked if this is possible.

**Comment by Al-Kindi-0** [comment]
Here is a more complete version of the above implementation. Note that, during absorption, the absorbed elements get added to the state as opposed to overwriting it. This is in order to make the argument for multiple-in-a-row squeezes sound since otherwise we would then have to overwrite the state with zeros each time we call a squeeze.

```rust
impl RpoRandomCoinDuplex {
    fn draw_basefield(&mut self) -> Felt {
        if self.output.is_empty() {
            Rpo256::apply_permutation(&mut self.state);
            self.output.extend_from_slice(&self.state[Rpo256::RATE_RANGE]);
        }
        self.output.pop().expect("Should not fail given the above")
    }
}

// A simplified version of the `SPONGE_PRG` reseedable pseudo-random number generator algorithm
// described in https://eprint.iacr.org/2011/499.pdf. The simplification is related to
// to the following facts:
// 1) A call to the reseed method implies one and only one call to the permutation function.
//  This is possible because in our case we never reseed with more than a rate number of elements. 
// 2) As a result of the previous point, we dont make use of input buffer to accumulate seed material.
impl RandomCoin for RpoRandomCoinDuplex {
    type BaseField = Felt;

    type Hasher = Rpo256;

    fn new(seed: &[Self::BaseField]) -> Self {
        let mut state = [ZERO; Rpo256::STATE_WIDTH];

        let digest = Rpo256::hash_elements(seed);
        let digest_elem = digest.as_elements();
        let rate_width = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
        let half_rate_width = rate_width / 2;

        for i in 0..half_rate_width {
            state[Rpo256::RATE_RANGE.start + i] += digest_elem[i];
        }

        // Absorb
        Rpo256::apply_permutation(&mut state);

        // Create output buffer
        let mut output = vec![];
        output.extend_from_slice(&state[Rpo256::RATE_RANGE]);

        RpoRandomCoinDuplex { state, output }
    }

    fn reseed(&mut self, data: <Self::Hasher as winter_crypto::Hasher>::Digest) {
        // Clear buffer
        self.output.clear();

        // Add the seed to the first half of the rate portion
        let data = data.as_elements();
        let rate_width = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
        let half_rate_width = rate_width / 2;
        for i in 0..half_rate_width {
            self.state[Rpo256::RATE_RANGE.start + i] += data[i];
        }

        // Absorb
        Rpo256::apply_permutation(&mut self.state);

        // Update output buffer
        self.output.extend_from_slice(&self.state[Rpo256::RATE_RANGE]);
    }

    fn reseed_with_int(&mut self, value: u64) {
        let value = Felt::new(value);
        self.state[Rpo256::RATE_RANGE.start] += value;
        Rpo256::apply_permutation(&mut self.state);
    }

    fn leading_zeros(&self) -> u32 {
        let seed_head = self.state[Rpo256::CAPACITY_RANGE.start].as_int();
        seed_head.trailing_zeros()
    }

    fn check_leading_zeros(&self, value: u64) -> u32 {
        let value = Felt::new(value);
        let mut state_tmp = [ZERO; Rpo256::STATE_WIDTH];
        
        for (i, s) in state_tmp.iter_mut().enumerate() {
            *s = self.state[i];
        }

        state_tmp[Rpo256::RATE_RANGE.start] += value;

        Rpo256::apply_permutation(&mut state_tmp);

        let seed_head = state_tmp[Rpo256::CAPACITY_RANGE.start].as_int();
        seed_head.trailing_zeros()
        
    }

    fn draw<E: FieldElement<BaseField = Felt>>(&mut self) -> Result<E, RandomCoinError> {
        let ext_degree = E::EXTENSION_DEGREE;
        let mut result = vec![];
        for i in 0..ext_degree {
            result[i] = self.draw_basefield();
        }
        let result = E::slice_from_base_elements(&result);
        Ok(result[0])
    }


    fn draw_integers(
        &mut self,
        num_values: usize,
        domain_size: usize,
    ) -> Result<Vec<usize>, RandomCoinError> {
        assert!(
            domain_size.is_power_of_two(),
            "domain size must be a power of two"
        );
        assert!(
            num_values < domain_size,
            "number of values must be smaller than domain size"
        );

        // determine how many bits are needed to represent valid values in the domain
        let v_mask = (domain_size - 1) as u64;

        // draw values from PRNG until we get as many unique values as specified by num_queries
        let mut values = Vec::new();
        for _ in 0..1000 {
            // get the next pseudo-random field element
            let value = self.draw_basefield().as_int();

            // use the mask to get a value within the range
            let value = (value & v_mask) as usize;

            if values.contains(&value) {
                continue;
            }
            values.push(value);
            if values.len() == num_values {
                break;
            }
        }

        if values.len() < num_values {
            return Err(RandomCoinError::FailedToDrawIntegers(
                num_values,
                values.len(),
                1000,
            ));
        }

        Ok(values)
    }
}

```

**Comment by bobbinth** [comment]
Looks great! Let's make a PR for this?

**Comment by Al-Kindi-0** [comment]
Sounds great!

**Comment by bobbinth** [comment]
Closed by #891.
